File: internal/services/layout_division.go
package services

const MaxTablesPerWaiter = 4

// FloorWaiterDivision is the suggested split of tables among floor staff.
// Coordination and the leader never receive tables. One waiter is always
// reserved as the couple's shadow. If the remaining waiters would exceed
// MaxTablesPerWaiter tables each, the co-leader can join the split.
type FloorWaiterDivision struct {
	TableCount       int
	WaiterCount      int
	CoordinatorCount int
	LeaderCount      int
	CoLeaderCount    int
	ShadowWaiters    int
	ServingWaiters   int
	UsedCoLeaders    int
	ServingPeople    int
	TablesPerPerson  float64
	OfferCoLeader    bool
	Shares           []int
}

func SuggestFloorWaiterDivision(tables, waiters, coordinators, leaders, coleaders int, includeCoLeader bool) FloorWaiterDivision {
	if tables < 0 {
		tables = 0
	}
	if waiters < 0 {
		waiters = 0
	}
	if coordinators < 0 {
		coordinators = 0
	}
	if leaders < 0 {
		leaders = 0
	}
	if coleaders < 0 {
		coleaders = 0
	}

	shadow := 0
	if waiters > 0 {
		shadow = 1
	}
	servingWaiters := waiters - shadow
	if servingWaiters < 0 {
		servingWaiters = 0
	}

	effectiveCoLeaders := coleaders
	if effectiveCoLeaders == 0 {
		effectiveCoLeaders = 1
	}
	offer := tables > 0 && (servingWaiters == 0 || tables > servingWaiters*MaxTablesPerWaiter)

	usedCo := 0
	if includeCoLeader && offer {
		usedCo = effectiveCoLeaders
	}

	serving := servingWaiters + usedCo
	per := 0.0
	if serving > 0 {
		per = float64(tables) / float64(serving)
	}

	shares := make([]int, 0, serving)
	if serving > 0 && tables > 0 {
		base := tables / serving
		extra := tables % serving
		for i := 0; i < serving; i++ {
			n := base
			if i < extra {
				n++
			}
			shares = append(shares, n)
		}
	}

	return FloorWaiterDivision{
		TableCount:       tables,
		WaiterCount:      waiters,
		CoordinatorCount: coordinators,
		LeaderCount:      leaders,
		CoLeaderCount:    coleaders,
		ShadowWaiters:    shadow,
		ServingWaiters:   servingWaiters,
		UsedCoLeaders:    usedCo,
		ServingPeople:    serving,
		TablesPerPerson:  per,
		OfferCoLeader:    offer,
		Shares:           shares,
	}
}

File: web/static/js/layout-division.js
(() => {
  const MAX_TABLES_PER_WAITER = 4;

  function suggestFloorWaiterDivision(input = {}) {
    const tables = Math.max(0, Number(input.tables) || 0);
    const waiters = Math.max(0, Number(input.waiters) || 0);
    const coordinators = Math.max(0, Number(input.coordinators) || 0);
    const leaders = Math.max(0, Number(input.leaders) || 0);
    const coleaders = Math.max(0, Number(input.coleaders) || 0);
    const includeCoLeader = Boolean(input.includeCoLeader);

    const shadowWaiters = waiters > 0 ? 1 : 0;
    const servingWaiters = Math.max(0, waiters - shadowWaiters);
    const effectiveCoLeaders = coleaders || 1;
    const offerCoLeader = tables > 0 && (servingWaiters === 0 || tables > servingWaiters * MAX_TABLES_PER_WAITER);
    const usedCoLeaders = includeCoLeader && offerCoLeader ? effectiveCoLeaders : 0;
    const servingPeople = servingWaiters + usedCoLeaders;
    const tablesPerPerson = servingPeople > 0 ? tables / servingPeople : 0;
    const shares = [];
    if (servingPeople > 0 && tables > 0) {
      const base = Math.floor(tables / servingPeople);
      const extra = tables % servingPeople;
      for (let index = 0; index < servingPeople; index += 1) {
        shares.push(base + (index < extra ? 1 : 0));
      }
    }

    return {
      tableCount: tables,
      waiterCount: waiters,
      coordinatorCount: coordinators,
      leaderCount: leaders,
      coLeaderCount: coleaders,
      shadowWaiters,
      servingWaiters,
      usedCoLeaders,
      servingPeople,
      tablesPerPerson,
      offerCoLeader,
      shares,
    };
  }

  function coleaderNames(count) {
    const total = Math.max(0, Number(count) || 0);
    if (total <= 0) return [];
    if (total === 1) return ["Colíder"];
    return Array.from({ length: total }, (_, index) => `Colíder ${index + 1}`);
  }

  window.emenysSuggestFloorWaiterDivision = suggestFloorWaiterDivision;
  window.emenysColeaderNames = coleaderNames;
})();
